lib/awful/placement.lua

```diff
@@ -702,23 +701,35 @@ local function get_relative_regions(geo, mode, is_absolute)
         end
     end
 
-    -- Get the drawable geometry
-    local dpos = geo.drawable and (
-        geo.drawable.drawable and
-            geo.drawable.drawable:geometry()
-            or geo.drawable:geometry()
-    ) or {x=0, y=0}
+    -- Get the parent geometry using one way or another depending on the object
+    -- Type
+    local bw, dgeo = 0, {x=0, y=0, width=1, height=1}
+
+    -- Detect various types of geometry table and (try) to get rid of the
+    -- differences so the code below don't have to care anymore.
+    if geo.drawin then
+        bw, dgeo = geo.drawin.border_width, geo.drawin:geometry()
+    elseif geo.drawable and geo.drawable.get_wibox then
+        bw   = geo.drawable.get_wibox().border_width
+        dgeo = geo.drawable.get_wibox():geometry()
+    elseif geo.drawable and geo.drawable.drawable then
+        bw, dgeo = 0, geo.drawable.drawable:geometry()
+    end
+
+    -- Add the infamous border size
+    dgeo.width  = dgeo.width  + 2*bw
+    dgeo.height = dgeo.height + 2*bw
 
     -- Compute the absolute widget geometry
-    local abs_widget_geo = is_absolute and geo or {
-        x            = dpos.x + geo.x              ,
-        y            = dpos.y + geo.y              ,
-        width        = geo.width                   ,
-        height       = geo.height                  ,
-        drawable     = geo.drawable                ,
+    local abs_widget_geo = is_absolute and dgeo or {
+        x            = dgeo.x + geo.x + bw,
+        y            = dgeo.y + geo.y + bw,
```

Comment:
It's subtle, but it's because it is a *widget* geometry inside the wibox being in the same code path as clients and wiboxes. I am not saying there isn't more "off by border width" bugs in that code, but that particular `+ bw` fixes one.

![screenshot_20181001_010328](https://user-images.githubusercontent.com/340384/46270506-ce3eb000-c515-11e8-9425-60885849a9f1.png)

Reply:
![screenshot_20181003_093037](https://user-images.githubusercontent.com/340384/46413469-fc2b1c80-c6ee-11e8-85cd-5c0cd9beb21b.png)

(unmodified from this PR with very large and colorful borders)

Reply:
Sorry, but what does this screenshot tell me exactly?
Is it "I added this code and afterwards the result looks correct"? If so: *Why* is adding the border width necessary here?

Reply:
if i understood correctly -- to make it aligned to the center of the border, not to the side of the border

Reply:
What do you mean with "center of the border"? To get the center of the border, I would expect x/y to be increased by `bw/2` and width/height to be decreased by `bw`.

Reply:
but we have 2*bw and bw

Reply:
Yeah, and I do not understand why we there are these values here.
Adding `2*bw` to the size and leaving the position as-is produces the geometry of the "outside" of a window. Adding `bw` to the position and leaving the size as-is produces the geometry of the "inside" of a window. Adding `2*bw` to size and adding `bw` to the position produces a geometry that is partly outside of the window (because the bottom right corner is moved by `3*bw`).

Reply:
`geo` does not contain the border, `dgeo` contain the border (line 723). This makes `geo` the content size. The content of `dgeo` starts at `x=bw, y=bw`.

The math is correct. This mode is to place a wibox next to a widget. Factoring the widget parent drawable border offset is correct.